File: Day230/motherVertex.go
package main

import "fmt"

func findMotherVertex(V int, adj [][]int) int {
	visited := make([]bool, V)
	mother := 0

	for i := 0; i < V; i++ {
		if !visited[i] {
			mother = i
			dfs(i, visited, adj)
		}
	}

	visited = make([]bool, V)
	dfs(mother, visited, adj)

	for i := 0; i < V; i++ {
		if !visited[i] {
			return -1
		}
	}

	return mother
}

func dfs(i int, visited []bool, adj [][]int) {
	visited[i] = true
	for _, n := range adj[i] {
		if !visited[n] {
			dfs(n, visited, adj)
		}
	}
}

func main() {
	V := 5
	edges := [][]int{{1, 0}, {0, 2}, {2, 1}, {2, 3}, {3, 4}}

	adj := make([][]int, V)
	for i := 0; i < V; i++ {
		adj[i] = []int{}
	}

	for i := 0; i < len(edges); i++ {
		adj[edges[i][0]] = append(adj[edges[i][0]], edges[i][1])
	}

	mother := findMotherVertex(V, adj)
	fmt.Println(mother)
}

File: Day230/motherVertex.java
import java.util.List;
import java.util.ArrayList;
public class motherVertex {
    public static List<Integer> convert(int[] arr) {
        List<Integer> list = new ArrayList<>();
        for (int i=0; i<arr.length; i++) {
            list.add(arr[i]);
        }
        return list;
    }
    public static void main(String[]args){
        int v = 5;
        int arr[][] = {{1,0},{0,2},{2,1},{2,3},{3,4}};
        List<List<Integer>>list = new  ArrayList<>();
        for (int i=0; i<arr.length; i++) { 
            List<Integer> temp = convert(arr[i]);
            list.add(temp);
        }
        System.out.println(list);
        System.out.println(findMotherVertex(v, list));
    }

    public static int findMotherVertex(int V, List<List<Integer>> adj) {

        boolean visited[] = new boolean[V];
        int mother = 0;
        for (int i=0; i<V; i++) {
            if (!visited[i]){
                mother = i;
                dfs(i, visited, adj);
            }
        }
        visited = new boolean[V];
        dfs(mother, visited, adj);
        for (int i=0; i<V; i++) {
            if (!visited[i]) {
                return -1;
            }
        }
        return mother;
    }
    public static void dfs(int i, boolean[] visited, List<List<Integer>> adj) {
        
        visited[i] = true;
        for (int n: adj.get(i)) {
            if (!visited[n]) {
                dfs(n, visited, adj);
            }
        }
        return;
    }
}
